.github/actions/python-maturin/action.yml

```diff
@@ -0,0 +1,236 @@
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+name: python-maturin
+description: Build/test/publish Python package with maturin
+inputs:
+  task:
+    description: "Task to run"
+    required: true
+  version:
+    description: "Version for publishing"
+    required: false
+    default: ""
+  dry_run:
+    description: "Dry run mode"
+    required: false
+    default: "false"
+
+runs:
+  using: "composite"
+  steps:
+    - name: Validate inputs
+      run: |
+        set -euo pipefail
+
+        # Validate task
+        case "${{ inputs.task }}" in
+          lint|test|build|publish)
+            echo "✓ Valid task: ${{ inputs.task }}"
+            ;;
+          *)
+            echo "❌ Invalid task: ${{ inputs.task }}"
+            exit 1
+            ;;
+        esac
+
+        # Validate version format if publishing
+        if [[ "${{ inputs.task }}" == "publish" && -n "${{ inputs.version }}" ]]; then
+          if ! echo "${{ inputs.version }}" | grep -qE '^[0-9]+\.[0-9]+\.[0-9]+'; then
+            echo "❌ Invalid version format: ${{ inputs.version }}"
+            exit 1
+          fi
+        fi
+      shell: bash
+
+    - name: Setup Python
+      uses: actions/setup-python@v5
+      with:
+        python-version: "3.11"
+
+    - name: Setup Rust with cache
+      uses: ./.github/actions/setup-rust-with-cache
+      with:
+        cache-targets: false  # Only cache registry and git deps, not target dir (sccache handles that)
+
+    - name: Cache pip
+      uses: actions/cache@v4
+      with:
+        path: ~/.cache/pip
+        key: pip-${{ runner.os }}-${{ hashFiles('foreign/python/pyproject.toml') }}
+
+    - name: Install dependencies
+      run: |
+        python -m pip install --upgrade pip wheel setuptools
+
+        # Install maturin for building
+        pip install 'maturin>=1.2,<2.0'
+
+        # Install testing dependencies from pyproject.toml
+        cd foreign/python
+        pip install -e ".[testing,examples,dev]"
+        cd ../..
+
+        # Install additional CI dependencies for Docker testing if needed
+        if [ "${{ inputs.task }}" = "test" ]; then
+          pip install 'testcontainers[docker]>=3.7.0,<5.0' || true
+        fi
+      shell: bash
+
+    - name: Lint and format check
+      if: inputs.task == 'lint'
+      run: |
+        # Build list of directories to check using absolute paths
+        DIRS_TO_CHECK="${GITHUB_WORKSPACE}/foreign/python/src"
+        DIRS_TO_CHECK="$DIRS_TO_CHECK ${GITHUB_WORKSPACE}/foreign/python/tests"
+        DIRS_TO_CHECK="$DIRS_TO_CHECK ${GITHUB_WORKSPACE}/foreign/python/python_examples"
+        DIRS_TO_CHECK="$DIRS_TO_CHECK ${GITHUB_WORKSPACE}/examples/python/getting-started"
+
+        echo "Directories to check: $DIRS_TO_CHECK"
+
+        echo "Running ruff check..."
+        ruff check $DIRS_TO_CHECK
```

Comment:
You can consider using isort with ruff:
```
ruff check --select I $DIRS_TO_CHECK
```

Note that `ruff check --select I` is slightly different from vanilla `isort`, so these two should not be used together.

I think it's better to create a script, e.g. `lint.sh`, for all these checks. Contributors can then run the script locally and fix trivial errors.